libs/experimental/langchain_experimental/llms/ollama_functions.py

```diff
@@ -280,6 +302,59 @@ class AnswerWithJustification(BaseModel):
         else:
             return llm | parser_chain
 
+    def _convert_messages_to_ollama_messages(
+        self, messages: List[BaseMessage]
+    ) -> List[Dict[str, Union[str, List[str]]]]:
+        ollama_messages: List = []
+        for message in messages:
+            role = ""
+            if isinstance(message, HumanMessage):
+                role = "user"
+            elif isinstance(message, AIMessage) or isinstance(message, ToolMessage):
+                role = "assistant"
+            elif isinstance(message, SystemMessage):
+                role = "system"
+            else:
+                raise ValueError("Received unsupported message type for Ollama.")
+
+            content = ""
+            images = []
+            if isinstance(message.content, str):
+                content = message.content
+            else:
+                for content_part in cast(List[Dict], message.content):
+                    if content_part.get("type") == "text":
+                        content += f"\n{content_part['text']}"
+                    elif content_part.get("type") == "image_url":
+                        if isinstance(content_part.get("image_url"), str):
```

Comment:
`_convert_messages_to_ollama_messages` in ChatOllama handles dict image_url, are we losing this functionality for OllamaFunctions?

Reply:
This is the same code from `ChatOllama` but has been tweaked to handle "ToolMessage" conversion to Ollama supported message payload.
Ollama only supports `system`, `user`, and `assistant` roles.
https://github.com/ollama/ollama/blob/main/docs/api.md#generate-a-chat-completion

The function is tweak to send `assistant` role for `ToolMessage` since Ollama doesn't support a `tool` or a `function` role yet.

https://github.com/lalanikarim/langchain/blob/1d5e66ce01da16c15341ddc9fdc4fc3a93078704/libs/experimental/langchain_experimental/llms/ollama_functions.py#L313

Reply:
有用。谢谢！